### PR TITLE
Scheduler Implementations V2

### PR DESCRIPTION
**Pull request by d80tb7**

### CODE REVIEW

File: internal/common/slices/slices.go [hunk 1]
```diff
@@ -7,3 +7,13 @@ func Map[T any, U any](list []T, fn func(val T) U) []U {
 	}
 	return out
 }
+
+func Filter[T any](list []T, fn func(val T) bool) []T {
```
Comment (severinson) [inline internal/common/slices/slices.go:11]: A nicer signature:
```
func Filter[S ~[]E, E any](s S, f func(v E) bool) S {
```

Comment (severinson) [reply internal/common/slices/slices.go:11]: This one works fort any alias of a slice too.

File: internal/armada/configuration/types.go [hunk 2]
```diff
@@ -172,6 +172,8 @@ type SchedulingConfig struct {
 	// All jobs in a gang must specify the total number of jobs in the gang via this annotation.
 	// The cardinality should be expressed as an integer, e.g., "3".
 	GangCardinalityAnnotation string
+	// If an executor hasn't heartbeated in this time period, it will be considered stale
+	StaleExecutorTimeout time.Duration
```
Comment (severinson) [inline internal/armada/configuration/types.go:176]:
```suggestion
	ExecutorTimeout time.Duration
```
I think this is clear enough.

File: internal/scheduler/adapters.go [hunk 3]
```diff
@@ -22,3 +24,31 @@ func (j *JobRepositoryAdapter) TryLeaseJobs(clusterId string, queue string, jobs
...
+
+type JobDbAdapter struct {
+	txn *memdb.Txn
+}
+
+func NewJobDbAdapter(txn *memdb.Txn) *JobDbAdapter {
+	return &JobDbAdapter{txn}
+}
+
+func (j *JobDbAdapter) GetJobIterator(_ context.Context, queue string) (JobIterator[*SchedulerJob], error) {
```
Comment (severinson) [inline internal/scheduler/adapters.go:36]: Unnecessary type parameter?

File: internal/scheduler/database/job_repository.go [hunk 4]
```diff
@@ -3,23 +3,164 @@ package database
...
+	queries := New(r.db)
+	count, err := queries.CountGroup(ctx, groupId)
+	if err != nil {
+		return 0, err
+	}
+	return uint32(count), nil
+}
+
+// fetch gets all rows from the database with a serial greater than from.
+// Rows are fetches in batches using the supplied fetchBatch function
```
Comment (severinson) [inline internal/scheduler/database/job_repository.go:151]:
```suggestion
// Rows are fetched in batches using the supplied fetchBatch function
```

File: internal/scheduler/reports.go [hunk 5]
```diff
@@ -159,7 +171,7 @@ func NewQueueSchedulingRoundReport[T LegacySchedulerJob](priorityFactor float64,
 	}
 }
 
-// Add a job scheduling report to the report for this invocation of the scheduler.
+// AddJobSchedulingReport Adds a job scheduling report to the report for this invocation of the scheduler.
```
Comment (severinson) [inline internal/scheduler/reports.go:174]:
```suggestion
// AddJobSchedulingReport adds a job scheduling report to the report for this invocation of the scheduler.
```

**Review by severinson — APPROVED** [review]

File: internal/scheduler/adapters.go (hunk 3 above)
Comment (d80tb7) [reply internal/scheduler/adapters.go:36]: I think it is needed- at least until we can remove the need for the scheduling report to contain a concrete job

**Review by severinson — APPROVED** [review]

**Review by severinson — APPROVED** [review]

**Pull request merged**